### PR TITLE
[BEAM-10527] Migrate Flink and Spark tests to pytest.

### PR DESCRIPTION
**Pull request by ibzib**
# Motivation

The main goals of migrating to pytest are:

1. Get Junit structured test output (BEAM-10527).
2. Replace PortableRunnerTest's bespoke timeout mechanism with pytest's (BEAM-9011). ~This also implicitly raises the timeout for all these tests from 60s to 600s, which should reduce the likelihood of timeout flakes (BEAM-8912).~

# Implementation

I originally wanted to do this in incremental changes, but I gradually realized a complete overhaul of these tests' configuration was needed. The main challenge was that `flink_runner_test.py` expected to be run as `__main__`, which is impossible with pytest. I basically reworked everything except the tests themselves; the tests themselves are unchanged.

## Test parametrization

- I left worker configuration in Gradle because it changes the test dependencies. I moved optimization and streaming into `flink_runner_test.py` because it removes the need to set up separate tox tasks, separate test result files, etc.
- Since `flink_job_server_driver`, `environment_type`, and `environment_config` are all pipeline options, I decided to pass them to `flink_runner_test.py` by introducing a global pytest option, `--test-pipeline-options`. `nose` uses `--test-pipeline-options` for integration tests, so I figured this would be generally useful beyond just these tests in the future.

# Bonus trivia

Prior to this change, we were running the exact same streaming test suite *four times* per Jenkins run.

Every `flinkCompatibilityMatrix` task ran the entirety of `flink_runner_test.py` which contained two classes: `FlinkRunnerTest` and `FlinkRunnerTestOptimized`. `FlinkRunnerTestOptimized` was basically the same thing as `FlinkRunnerTest`, but it added the `pre_optimize=all` experiment and skipped external transform tests, since the Python optimizer breaks external transforms (BEAM-7252). But we were *also* adding `pre_optimize=all` in Gradle, redundantly.

The old configuration looks like this:

```groovy
dependsOn flinkCompatibilityMatrix(streaming: false, workerType: CompatibilityMatrixConfig.SDK_WORKER_TYPE.LOOPBACK)
dependsOn flinkCompatibilityMatrix(streaming: true, workerType: CompatibilityMatrixConfig.SDK_WORKER_TYPE.LOOPBACK)
dependsOn flinkCompatibilityMatrix(streaming: true, workerType: CompatibilityMatrixConfig.SDK_WORKER_TYPE.LOOPBACK, preOptimize: true)
```

Notice that pre-optimized batch is missing. This is because `flinkCompatibilityMatrixBatchPreOptimize*` would run `FlinkRunnerTest` with `pre_optimize=all` but without skipping the external transform tests, causing failure.

What about streaming, then? Well, the optimizer *doesn't affect streaming pipelines at all*: https://github.com/apache/beam/blob/489cf2cbf335f372060469953ed7599f2838a591/sdks/python/apache_beam/runners/portability/portable_runner.py#L319

So in one invocation of `flinkValidatesRunner`, `flinkCompatibilityMatrixStreamingLoopback` would run `FlinkRunnerTest` (without `pre_optimize=all`) and `FlinkRunnerTestOptimized`, then `flinkCompatibilityMatrixStreamingPreOptimizeLoopback` would run `FlinkRunnerTest` (with `pre_optimize=all`) and `FlinkRunnerTestOptimized` (with `pre_optimize=all` twice). Besides the skips in `FlinkRunnerTestOptimized`, all four tests would be doing the exact same thing.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ibzib** [comment]
Run Portable_Python PreCommit

**Comment by ibzib** [comment]
R: @udim @mxm

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py [hunk 1]
```diff
@@ -53,361 +53,380 @@
...
+  def __init__(self, *args, **kwargs):
+    super(FlinkRunnerTest, self).__init__(*args, **kwargs)
+    self.environment_type = None
+    self.environment_config = None
+
+  @pytest.fixture(autouse=True)
+  def parse_options(self, request):
+    test_pipeline_options = (
+        request.config.option.test_pipeline_options
+        if request.config.option.test_pipeline_options else '')
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/flink_runner_test.py:86]: nit:
```suggestion
        request.config.option.test_pipeline_options or '')
```

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py [hunk 2]
```diff
@@ -53,361 +53,380 @@
...
+    parser.add_argument(
+        '--environment_config',
+        help='Set environment configuration for running the user code.\n For '
+        'DOCKER: Url for the docker image.\n For PROCESS: json of the '
+        'form {"os": "<OS>", "arch": "<ARCHITECTURE>", "command": '
+        '"<process to execute>", "env":{"<Environment variables 1>": '
+        '"<ENV_VAL>"} }. All fields in the json are optional except '
+        'command.')
+    known_args, unknown_args = parser.parse_known_args(
+        test_pipeline_options.split())
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/flink_runner_test.py:109]: Could this be problematic if one of the option values contains a space, e.g. as part of a json string for the environment config?

File: sdks/python/apache_beam/runners/portability/spark_runner_test.py [hunk 3]
```diff
@@ -21,141 +21,158 @@
...
+    parser.add_argument(
+        '--environment_config',
+        help='Set environment configuration for running the user code.\n For '
+        'DOCKER: Url for the docker image.\n For PROCESS: json of the '
+        'form {"os": "<OS>", "arch": "<ARCHITECTURE>", "command": '
+        '"<process to execute>", "env":{"<Environment variables 1>": '
+        '"<ENV_VAL>"} }. All fields in the json are optional except '
+        'command.')
+    known_args, unknown_args = parser.parse_known_args(
+        test_pipeline_options.split())
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/spark_runner_test.py:79]: Same here, what about whitespace in the values?

**Comment by mxm** [comment]
Run Java PreCommit

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py (hunk 1 above)
Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/flink_runner_test.py]: done

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py (hunk 2 above)
Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/flink_runner_test.py]: Good catch. I used `shlex.split(...)` instead.

**Comment by ibzib** [comment]
> I couldn't find the published test results in Jenkins. Do we have to add this separately? https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Commit/6126/

This is because publishing the test results relies on a change to the Jenkins job config, and I didn't want to `run seed job` because Flink PVR is a precommit and if I made a mistake, `run seed job` could break it for others. I tried testing locally on dockerized Jenkins, but I couldn't get it to work.

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py (hunk 2 above)
Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/flink_runner_test.py]: I also had to make many changes to preserve the correct string quoting. Tox in particular does something weird with quotes in its commands, so I had to double-escape the strings before passing to Tox and unescaping them after.

File: sdks/python/apache_beam/runners/portability/spark_runner_test.py (hunk 3 above)
Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/spark_runner_test.py]: done

**Comment by ibzib** [comment]
I got `flinkCompatibilityMatrixPROCESS` to pass on my machine by escaping the arguments via `${1@Q}`. Apparently whatever shell Jenkins is using does not support this. I will have to find a better solution.

I have already spent a long time trying to fix quotes, so I can't help but wondering: why do we need `flinkCompatibilityMatrixPROCESS` in the first place, when it is not being run anywhere? If it's important, shouldn't we add it to some postcommit?

**Comment by ibzib** [comment]
> I have already spent a long time trying to fix quotes, so I can't help but wondering: why do we need flinkCompatibilityMatrixPROCESS in the first place, when it is not being run anywhere? If it's important, shouldn't we add it to some postcommit?

Another solution I had in mind was reworking the `--environment_config` option. JSON blobs are unwieldy, and overloading the `--environment_config` option is confusing to the user. We could make each field in the PROCESS `--environment_config` blob its own argument, and then reject these arguments when `environment_type != PROCESS`.

File: sdks/python/scripts/run_pytest.sh [hunk 4]
```diff
@@ -29,10 +29,16 @@ posargs=$2
 
 # Run with pytest-xdist and without.
 pytest -o junit_suite_name=${envname} \
-  --junitxml=pytest_${envname}.xml -m 'not no_xdist' -n 6 --pyargs ${posargs}
+  --junitxml=pytest_${envname}.xml -m 'not no_xdist' -n 6 \
+  --ignore=apache_beam/runners/portability/flink_runner_test.py \
```
Comment (udim) [inline sdks/python/scripts/run_pytest.sh:33]: Can you instead check if `test-pipeline-options` is set? I believe that's how all the other ITs are skipped (see `TestPipeline` code).

**Comment by mxm** [comment]
Thanks for fixing the quoting issue!

> I got `flinkCompatibilityMatrixPROCESS` to pass on my machine by escaping the arguments via `${1@Q}`. Apparently whatever shell Jenkins is using does not support this. I will have to find a better solution.

I think that feature only works in newer versions of `bash`. 

> why do we need `flinkCompatibilityMatrixPROCESS` in the first place, when it is not being run anywhere? If it's important, shouldn't we add it to some postcommit?

`flinkCompatibilityMatrixPROCESS` was how we ran the PVR tests to avoid a dependency on the container build (for speed). I was under the assumption that we are still doing that. That probably changed when we added the external transform tests which require the Java container. I agree that we should at least have a post commit.

**Comment by ibzib** [comment]
> > I have already spent a long time trying to fix quotes, so I can't help but wondering: why do we need flinkCompatibilityMatrixPROCESS in the first place, when it is not being run anywhere? If it's important, shouldn't we add it to some postcommit?
> 
> Another solution I had in mind was reworking the `--environment_config` option. JSON blobs are unwieldy, and overloading the `--environment_config` option is confusing to the user. We could make each field in the PROCESS `--environment_config` blob its own argument, and then reject these arguments when `environment_type != PROCESS`.

@mxm what do you think about https://issues.apache.org/jira/browse/BEAM-10671?

**Comment by mxm** [comment]
> > > I have already spent a long time trying to fix quotes, so I can't help but wondering: why do we need flinkCompatibilityMatrixPROCESS in the first place, when it is not being run anywhere? If it's important, shouldn't we add it to some postcommit?
> > 
> > 
> > Another solution I had in mind was reworking the `--environment_config` option. JSON blobs are unwieldy, and overloading the `--environment_config` option is confusing to the user. We could make each field in the PROCESS `--environment_config` blob its own argument, and then reject these arguments when `environment_type != PROCESS`.
> 
> @mxm what do you think about https://issues.apache.org/jira/browse/BEAM-10671?

I think it makes sense. Especially for error reporting by having dedicated argument parsers for all environments parameters.

**Comment by ibzib** [comment]
Run Python Spark ValidatesRunner

File: sdks/python/scripts/run_pytest.sh (hunk 4 above)
Comment (ibzib) [reply sdks/python/scripts/run_pytest.sh]: Done

**Comment by mxm** [comment]
This looks good to me after rebasing. Thanks for porting this to pytest!

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py [hunk 5]
```diff
@@ -53,361 +54,386 @@
 from apache_beam.transforms import userstate
 from apache_beam.transforms.sql import SqlTransform
 
+# Run as
+#
+# pytest flink_runner_test.py \
+#     [--test_pipeline_options "--flink_job_server_jar=/path/to/job_server.jar \
```
Comment (boyuanzz) [inline sdks/python/apache_beam/runners/portability/flink_runner_test.py:60]: --test-pipeline-options

Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/flink_runner_test.py:60]: Oh yeah, `--test_pipeline_options` is now required (even though it should be possible to leave it empty). Boyuan, would you mind filing a PR to fix this?

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py [hunk 6]
```diff
@@ -53,361 +54,386 @@
...
+#                               --environment_type=DOCKER"] \
+#     [FlinkRunnerTest.test_method, ...]
```
Comment (boyuanzz) [inline sdks/python/apache_beam/runners/portability/flink_runner_test.py:62]: And the test filter here doesn't work for me properly. The working version for me is 
"""
pytest flink_runner_test.py::TestClass:test_case --test-pipeline-options "--flink_job_server_jar=XXX --environment_type=XXX "
"""

**Pull request merged**